### ISSUE TITLE
Esp32 BLE+ETH+OTA

### ISSUE THREAD

**Issue by robcim** [espressif/arduino-esp32#2962]
Hi all,
I'm Roberto. I use a ESP32 ( flash size 16MB) with BLE and ETH communication. I add the OTA mode (WebUpdate) but he's not work. 
I have this error in console: [E][WiFiClient.cpp:338] write(): 11

Anyone can help me please.
Thank you

my code is:

#include <WiFi.h>
#include <WiFiClient.h>
#include <WebServer.h>
#include <ESPmDNS.h>
#include <Update.h>
#include <ETH.h>

#include <HardwareSerial.h>
HardwareSerial Serial485(2);

int porta_stato_check = 0;
static bool eth_connected = false;
bool deviceConnected = false;
bool oldDeviceConnected = false;

#define ETH_CLK_MODE    ETH_CLOCK_GPIO0_IN   // nostre board ETH_CLOCK_GPIO0_IN      //  ETH_CLOCK_GPIO17_OUT
#define ETH_POWER_PIN   17                     
#define ETH_TYPE        ETH_PHY_LAN8720  
#define ETH_ADDR        1                     
#define ETH_MDC_PIN     23
#define ETH_MDIO_PIN    18

WebServer server(80);

static const char serverIndex[] PROGMEM = R"(
<!DOCTYPE html>
<html>
 <head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1,user-scalable=no">
      <title>Upgrade</title>
      <style>div,fieldset,input,select{padding:10px;font-size:1em;}
        fieldset{background-color:yellow;}
        p{margin:0.5em 0;button:hover{background-color:#32CD32;}.bred{background-color:green;}}
        input{width:100%;box-sizing:border-box;-webkit-box-sizing:border-box;-moz-box-sizing:border-box;button:hover{background-color:#32CD32;}.bred{background-color:green;}}
        input[type=checkbox],input[type=radio]{width:1em;margin-right:6px;vertical-align:-1px;}
        select{width:100%;}
        textarea{resize:none;width:98%;height:318px;padding:5px;overflow:auto;}
        body{text-align:center;font-family:verdana;font-size:1.6rem;}
        td{padding:0px;}
        button{border:0;border-radius:0.3rem;background-color:#00FF00;color:red;line-height:2.4rem;font-size:1.2rem;width:100%;-webkit-transition-duration:0.4s;transition-duration:0.4s;cursor:pointer;}     
      </style>
  <body>    
  <div style="text-align:left;display:newline-block;min-width:300px;">
    <div style="text-align:center;">
      <h2>Upgrade Firmware</h2></div>
    <form method='POST' action='/update' enctype='multipart/form-data'>
      <p></p>
      <p></p>
      <input type='file' name='update'>
      <p></p>
      <p></p>
      <input type='submit' value='Update'>
    </form>
      <p></p>
      <br>
      <p></p>
      <br>  
    <form action="." method="get">
      <button style=" background-color: #00FF00;color: red;">Main Menu</button>
    </form><p></p>
  </body>
          <div style="text-align:right;font-size:15px;"><hr>
          <a href="" target="_blank" style="color:#000000;">Smart Device</a>
          </div></div>
  </head>
</html>
)";

void OTA() {

  const char* host_ = "prova04072019";    
  Serial.print("host OTA= "); Serial.println(host_);
  MDNS.begin(host_);

  server.on("/up", HTTP_GET, []() {   // "/up"
    server.sendHeader("Connection", "close");
    server.send(200, "text/html", serverIndex);
  });
  server.on("/update", HTTP_POST, []() {
    server.sendHeader("Connection", "close");
    server.send(200, "text/plain", (Update.hasError()) ? "FAIL" : "OK");
    ESP.restart();
  }, []() {
    HTTPUpload& upload = server.upload();
    if (upload.status == UPLOAD_FILE_START) {
      Serial.setDebugOutput(true);
      Serial.printf("Update: %s\n", upload.filename.c_str());
      if (!Update.begin()) { //start with max available size
        Update.printError(Serial);
      }
    } else if (upload.status == UPLOAD_FILE_WRITE) {
      if (Update.write(upload.buf, upload.currentSize) != upload.currentSize) {
        Update.printError(Serial);
      }
    } else if (upload.status == UPLOAD_FILE_END) {
      if (Update.end(true)) { //true to set the size to the current progress
        Serial.printf("Update Success: %u\nRebooting...\n", upload.totalSize);
      } else {
        Update.printError(Serial);
      }
      Serial.setDebugOutput(false);
    }
  });
//  server.on("/cn", HTTP_GET, []() {
//    server.sendHeader("Connection", "close");
//    server.send(200, "text/html", serverConf);
//  });
//  server.on("/wi", HTTP_GET, []() {
//    server.sendHeader("Connection", "close");
//    server.send(200, "text/html", serverWifi);
//  });
//  server.on("/rst", HTTP_GET, []() {
//    //      server.sendHeader("Connection", "close");
//    //      server.send(200, "text/html", serverWifi);
//    delay(500);
//    ESP.restart();
//  });
//  server.on("/mq", HTTP_GET, []() {
//    server.sendHeader("Connection", "close");
//    server.send(200, "text/html", serverMqtt);
//    String mh = server.arg("mh");
//    String md = server.arg("md");
//    String mu = server.arg("mu");
//    String mp = server.arg("mp");
//    String mt = server.arg("mt");
//    if (mh.length() > 0 && md.length() > 0 && mu.length() > 0 && mp.length() > 0 && mt.length() > 0 ) {
//      Serial.print("server_mqtt= "); Serial.println(mh);
//      Serial.print("port_mqtt= "); Serial.println(md);
//      Serial.print("user_mqtt= "); Serial.println(mu);
//      Serial.print("pasw_mqtt= "); Serial.println(mp);
//      Serial.print("topic= "); Serial.println(mt);
//    }
//  });
//  server.on("/wi_dhcp", HTTP_GET, []() {
//    server.sendHeader("Connection", "close");
//    server.send(200, "text/html", serverWifiDHCP);
//    String s1 = server.arg("s1");
//    String p1 = server.arg("p1");
//    if (s1.length() > 0 && p1.length() > 0) {
//      Serial.print("ssid= "); Serial.println(s1);
//      Serial.print("password= "); Serial.println(p1);
//    }
//  });
//  server.on("/wi_static", HTTP_GET, []() {
//    server.sendHeader("Connection", "close");
//    server.send(200, "text/html", serverWifiStatic);
//    String s1 = server.arg("s1");
//    String p1 = server.arg("p1");
//    String ip = server.arg("ip");
//    String msk = server.arg("msk");
//    String gw = server.arg("gw");
//    String dn = server.arg("dn");
//    if (s1.length() > 0 && p1.length() > 0 && ip.length() > 0 && msk.length() > 0 && gw.length() > 0 && dn.length() > 0) {
//      Serial.print("ssid= "); Serial.println(s1);
//      Serial.print("password= "); Serial.println(p1);
//      Serial.print("ip_add= "); Serial.println(ip);
//      Serial.print("mask= "); Serial.println(msk);
//      Serial.print("gateway= "); Serial.println(gw);
//      Serial.print("dns= "); Serial.println(dn);
//    }
//  });
  server.begin();
  MDNS.addService("http", "tcp", 80);
}

void setup(void) {
  Serial.begin(115200);
  Serial.println();
  Serial.println("Booting Sketch...");

  ETH.begin(ETH_ADDR, ETH_POWER_PIN, ETH_MDC_PIN, ETH_MDIO_PIN, ETH_TYPE, ETH_CLK_MODE);
  
  OTA();
  
}

void loop(void) {
  server.handleClient();
  delay(1);
}

**Comment by lbernstone** [comment]
tl;dr
Please [format your code](https://help.github.com/articles/creating-and-highlighting-code-blocks/), cut out the huge chunk of comment, and follow the [issue template](https://github.com/espressif/arduino-esp32/blob/master/.github/ISSUE_TEMPLATE/bug_report.md).  Provide as much logging as possible- a single error line with no context is not useful.

**Comment by robcim** [comment]
#include <WiFi.h>
#include <WiFiClient.h>
#include <WebServer.h>
#include <ESPmDNS.h>
#include <Update.h>
#include <ETH.h>

#include <HardwareSerial.h>
HardwareSerial Serial485(2);

int porta_stato_check = 0;
static bool eth_connected = false;
bool deviceConnected = false;
bool oldDeviceConnected = false;

#define ETH_CLK_MODE ETH_CLOCK_GPIO0_IN
#define ETH_POWER_PIN 17
#define ETH_TYPE ETH_PHY_LAN8720
#define ETH_ADDR 1
#define ETH_MDC_PIN 23
#define ETH_MDIO_PIN 18

WebServer server(80);

static const char serverIndex[] PROGMEM = R"(

<title>Upgrade</title> 
<style>div,fieldset,input,select{padding:10px;font-size:1em;} fieldset{background-color:yellow;} 
    p{margin:0.5em 0;button:hover{background-color:#32CD32;}.bred{background-color:green;}} 
    input{width:100%;box-sizing:border-box;-webkit-box-sizing:border-box;-moz-box-sizing:border- 
    box;button:hover{background-color:#32CD32;}.bred{background-color:green;}} 
    input[type=checkbox],input[type=radio]{width:1em;margin-right:6px;vertical-align:-1px;} 
    select{width:100%;} textarea{resize:none;width:98%;height:318px;padding:5px;overflow:auto;} 
    body{text-align:center;font-family:verdana;font-size:1.6rem;} td{padding:0px;} button{border:0;border- 
    radius:0.3rem;background-color:#00FF00;color:red;line-height:2.4rem;font-size:1.2rem;width:100%;- 
    webkit-transition-duration:0.4s;transition-duration:0.4s;cursor:pointer;} 
</style>
  <body>    
  <div style="text-align:left;display:newline-block;min-width:300px;">
    <div style="text-align:center;">
      <h2>Upgrade Firmware</h2></div>
    <form method='POST' action='/update' enctype='multipart/form-data'>
      <input type='file' name='update'>
      <input type='submit' value='Update'>
    </form>
    <form action="." method="get">
      <button style=" background-color: #00FF00;color: red;">Main Menu</button>
    </form><p></p>
  </body>
          <div style="text-align:right;font-size:15px;"><hr>
          <a href="" target="_blank" style="color:#000000;">Smart Device</a>
          </div></div>
  </head>
</html>
)";

void WiFiEvent(WiFiEvent_t event)
{
  switch (event) {
    case SYSTEM_EVENT_ETH_START:
      Serial.println("ETH Started");
      //set eth hostname here
      ETH.setHostname("esp32-ethernet");
      break;
    case SYSTEM_EVENT_ETH_CONNECTED:
      Serial.println("ETH Connected");
      break;
    case SYSTEM_EVENT_ETH_GOT_IP:
      Serial.print("ETH MAC: ");
      Serial.print(ETH.macAddress());
      Serial.print(", IPv4: ");
      Serial.print(ETH.localIP());
      if (ETH.fullDuplex()) {
        Serial.print(", FULL_DUPLEX");
      }
      Serial.print(", ");
      Serial.print(ETH.linkSpeed());
      Serial.println("Mbps");
      eth_connected = true;
      break;
    case SYSTEM_EVENT_ETH_DISCONNECTED:
      Serial.println("ETH Disconnected");
      eth_connected = false;
      break;
    case SYSTEM_EVENT_ETH_STOP:
      Serial.println("ETH Stopped");
      eth_connected = false;
      break;
    default:
      break;
  }
}

void testClient(const char * host, uint16_t port)
{
  Serial.print("\nconnecting to ");
  Serial.println(host);

  WiFiClient client;
  if (!client.connect(host, port)) {
    Serial.println("connection failed");
    return;
  }
  client.printf("GET / HTTP/1.1\r\nHost: %s\r\n\r\n", host);
  while (client.connected() && !client.available());
  while (client.available()) {
    Serial.write(client.read());
  }

  Serial.println("closing connection\n");
  client.stop();
}

class MyCallbacks: public BLECharacteristicCallbacks {
    void onWrite(BLECharacteristic *pCharacteristic) {
      std::string value = pCharacteristic->getValue();
      if (value.length() > 0) {
        Serial.println("*********"); Serial.println("New pass arrived: ");
        for (int i = 0; i < value.length(); i++)
          Serial.print(value[i]);
      }
    }
};
class MyServerCallbacks: public BLEServerCallbacks {
    void onConnect(BLEServer* pServer) {
      deviceConnected = true;
      Serial.println("CB connesso true");
    };
    void onDisconnect(BLEServer* pServer) {
      deviceConnected = false;
      Serial.println("CB non connesso true");
    }
};

void OTA() {

   const char* host_ = "prova04072019";
   Serial.print("host OTA= "); Serial.println(host_);
   MDNS.begin(host_);

   server.on("/up", HTTP_GET, { // "/up"
   server.sendHeader("Connection", "close");
   server.send(200, "text/html", serverIndex);
});
   server.on("/update", HTTP_POST, {
   server.sendHeader("Connection", "close");
   server.send(200, "text/plain", (Update.hasError()) ? "FAIL" : "OK");
   ESP.restart();
}, {
HTTPUpload& upload = server.upload();
if (upload.status == UPLOAD_FILE_START) {
   Serial.setDebugOutput(true);
   Serial.printf("Update: %s\n", upload.filename.c_str());
if (!Update.begin()) { //start with max available size
   Update.printError(Serial);
}
} else if (upload.status == UPLOAD_FILE_WRITE) {
if (Update.write(upload.buf, upload.currentSize) != upload.currentSize) {
   Update.printError(Serial);
}
} else if (upload.status == UPLOAD_FILE_END) {
if (Update.end(true)) { 
   Serial.printf("Update Success: %u\nRebooting...\n", upload.totalSize);
} 
else {
   Update.printError(Serial);
}
 Serial.setDebugOutput(false);
}
});

   server.begin();
   MDNS.addService("http", "tcp", 80);
}

void init_ble() {

  BLEDevice::init("TFC_");   // (clientId.c_str());

  pServer = BLEDevice::createServer();
  pServer->setCallbacks(new MyServerCallbacks());
  init_service_charact();

  pCharacteristicStart->setValue(" ");

  pCharacteristicP1->setValue(" ");
  pCharacteristicP2->setValue(" ");
  pCharacteristicOpen ->setValue(" ");

  pCharacteristicIDCall->setValue(" ");
  pCharacteristicIDInt->setValue(" ");
  pCharacteristicIDPsw->setValue(" ");

  pCharacteristicP1->setCallbacks(new MyCallbacks());
  pCharacteristicP2->setCallbacks(new MyCallbacks());
  pCharacteristicOpen->setCallbacks(new MyCallbacks());

  pServiceLogin->start();
  pServiceStart->start();
  pServiceStatus->start();
  pServiceActions->start();

  pServer->getAdvertising()->start();
}

void read_statusBOX () {

  Serial.print("*****read_statusBOX*****"); 
  pCharacteristicStatus->setValue("1");
  pCharacteristicFailed->setValue("3");
}

void setup(void) {
   Serial.begin(115200);
   Serial.println();
   Serial.println("Booting Sketch...");
   
   WiFi.onEvent(WiFiEvent);
   
   ETH.begin(ETH_ADDR, ETH_POWER_PIN, ETH_MDC_PIN, ETH_MDIO_PIN, ETH_TYPE, ETH_CLK_MODE);
   
   init_ble();

   OTA();
}

void loop(void) {
  if (eth_connected) {
    testClient("google.com", 80);
  }
  if (deviceConnected) {
    delay(100);
    read_statusBOX();
  }
  else {
    Serial.println("BLE not connected");

    pCharacteristicStart->setValue("");
    pCharacteristicP1->setValue(" ");
    pCharacteristicP2->setValue(" ");
    int check_sens_porta = controllo_porta(); // digitalRead(contatto_porta); 

    if (check_sens_porta == porta_open) {
      statoled_allarme = 1;
      pCharacteristicOpen->setValue("3");

      Serial.println("non connesso BLE, allarme porta aperta");
    }
  }
   server.handleClient();
   delay(1);
}
